### PR TITLE
Add ONNX export support for huggingface's bigscience/bloom-560m model

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Port fix from https://github.com/huggingface/safetensors/pull/318 into ONNX exporter until it is merged

* This add support for safetensors to be loaded within a FakeTensorMode, which results in creating `torch.empty((shape,), dtype=)`. This is done through a monkeypatch for the in-progress https://github.com/huggingface/safetensors/pull/318
* Adds a test for the HF bloom model (bigscience/bloom-560m)
* This PR also fixes existing fake tensor unit tests by moving the `torch.onnx.dynamo_export` to be inside the `enable_fake_mode()` context. Although calling `torch.onnx._dynamo_export` works for several models, the right way of using fake mode is calling the exporter within the context manager.

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
@BowenBao @justinchuby @titaiwangms @wschin Does anyone know why the `safetensors` patching does not work when the bloom model is ran within pytest, but works fine on a standalone script?

**Review by BowenBao — APPROVED** [review]
Does safetensor route work with the torch.load patch to provide `export_output.save` with path to real model states later?

File: torch/onnx/_internal/fx/patcher.py [hunk 1]
```diff
@@ -77,6 +91,33 @@ def _rebuild_real_tensor(storage, storage_offset, size, stride):
         self.torch_load_wrapper = torch_load_wrapper
         self.torch__util_rebuild_tensor_wrapper = torch__util__rebuild_tensor_wrapper
 
+        if has_safetensors_and_transformers:
+
+            def safetensors_load_file_wrapper(filename, device="cpu"):
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/patcher.py:96]: We can follow up with adjustment according to https://github.com/pytorch/pytorch/issues/106732.
I'm ok with this PR as the patch to unblock.

**Comment by thiagocrepaldi** [comment]
Please don't merge this yet. Trying to see if I can come up with a better change in a couple hours. Will also address comments

**Comment by titaiwangms** [comment]
> Please don't merge this yet. Trying to see if I can come up with a better change in a couple hours. Will also address comments

It's worth trying to update the transformers version in this PR to see if it works on a newer version of transformers. We will have to update it after that one is merged anyway.

**Comment by thiagocrepaldi** [comment]
> Does safetensor route work with the torch.load patch to provide `export_output.save` with path to real model states later?

No, that is why this test is within `test_fx_to_onnx.py` only.

We can do this by loading the whole state_dict in memory (and that is an ok strategy for now), but we need to find a way to convert a safetensors into pytorch compatible state_dict out-of-memory, probably by reading tensors by tensor from the safetensors and writing it on another pytorch format ckp

**Comment by BowenBao** [comment]
I wonder if it is possible to lazy process that at `export_output.save`, stream loading safetensors and saving to ONNX format.

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 2]
```diff
@@ -998,6 +998,40 @@ def create_kwargs():
...
+        "ValueError: Message onnx.ModelProto exceeds maximum protobuf size of 2GB"
+        "[ONNXRuntimeError] : Status Message: Indices vs updates dimensions differs at position=1 0 vs 3"
+    )
+    @pytorch_test_common.skip_dynamic_fx_test(
+        "RuntimeError:: SymIntArrayRef expected to contain only concrete integers"
+    )
+    @pytorch_test_common.skip_load_checkpoint_after_model_creation(
+        "HF Bloom model does not need `model.load_state_dict` to work."
+    )
+    def test_fake_tensor_mode_huggingface_bigscience__bloom_560m(self):
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:1011]:
```suggestion
    def test_fake_tensor_mode_huggingface_bigscience_bloom_560m(self):
```

File: test/onnx/test_fx_to_onnx.py [hunk 3]
```diff
@@ -351,34 +350,50 @@ def forward(self, x):
...
-            _ = torch.onnx.dynamo_export(
-                fake_model, real_x, export_options=export_options
+            # Scenario 4: Fake model and real input WITH fake_context
+            with self.assertRaises(torch.onnx.OnnxExporterError):
+                export_options = ExportOptions(fake_context=fake_context)
+                _ = torch.onnx.dynamo_export(
+                    fake_model, real_x, export_options=export_options
+                )
+
+    def test_fake_tensor_mode_huggingface_bigscience__bloom_560m(self):
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx.py:382]: Do we need the same model in both runtime/no_runtime tests?

Comment (thiagocrepaldi) [reply test/onnx/test_fx_to_onnx.py:382]: when the one with ORT passes, the other one can be deleted.

**Pull request closed without merging**